diff --git a/lcl/interfaces/win32/win32callback.inc b/lcl/interfaces/win32/win32callback.inc
index 2f58313736..01e659bc53 100644
--- a/lcl/interfaces/win32/win32callback.inc
+++ b/lcl/interfaces/win32/win32callback.inc
@@ -1440,6 +1440,18 @@ begin
       Area := PDrawIS^.rcItem;
       ItemState := TOwnerDrawState(PDrawIS^.itemState);
       DC := PDrawIS^._hDC;
+      // odFocused on enter control didn't fired with ODA_DRAWENTIRE
+      if (PDrawIS^.itemAction and ODA_DRAWENTIRE)<>0 then
+      begin
+        if lWinControl.Focused and not (odFocused in ItemState) then
+           if ((PDrawIS^.CtlType=ODT_LISTBOX)
+               and (TCustomListBox(lWinControl).ItemIndex=ItemID)) or
+              ((PDrawIS^.CtlType=ODT_LISTVIEW)
+               and (TCustomListView(lWinControl).ItemIndex=ItemID)) or
+              ((PDrawIS^.CtlType=ODT_COMBOBOX)
+               and (TCustomComboBox(lWinControl).ItemIndex=ItemID)) then
+                  Include(ItemState,odFocused);
+      end;
     end;
     if (aMsg = LM_DRAWLISTITEM) and (WindowInfo <> @DefaultWindowInfo) then
     begin
